techlibs/gowin/synth_gowin.cc

```diff
@@ -112,6 +122,14 @@ struct SynthGowinPass : public ScriptPass
 				vout_file = args[++argidx];
 				continue;
 			}
+			if (args[argidx] == "-json" && argidx+1 < args.size()) {
+				json_file = args[++argidx];
+				nobram = true;
+				nolutram = true;
+				nowidelut = true;
+				noalu = true;
```

Comment:
Better do not add these 4 lines. That way yosys part should not be changed while you are improving nextpnr support. Yes command line will be longer, but you can explicitly disable/enable certain flags then.

Reply:
Hmm, I would like `synth_gowin` to do the right thing as much as possible without people having to know all the internal details of supported Gowin primitives. But on the other hand... yea it's be a new PR every time I support more primitives.

Reply:
With ECP5 I added support for all primitives in Yosys by default before they were supported in nextpnr. In retrospect, this had the downside that people were including the Yosys args to disable those primitives in their own code and not removing them when those primitives were supported in nextpnr so getting worse results than needed. So I don't know which the best way is, tbh.

Reply:
Yea, I would be somewhat in favour of a plain `synth_gowin` to follow "current best practice" as much as possible. It might be a long time between `nextpnr-gowin` becoming usable and it supporting all the primitives. Removing those lines later should be a no-brainer.

Reply:
I agree with Daves point, but with flags as they are now, you can not enable alu for example when outputting json. From user perspective current code is better, from developer perspective you would need to recompile yosys when you add new thing in nextpnr, and basically your nextpnr version will work only with latest yosys in that case. So yeah hard decision.

Reply:
If you just want a JSON with all the options you can just not pass the `-json` option and issue a simple `write_json` command. Maybe the `-json` option can get a name that indicates it's specifically for Nextpnr/Apicula.

Reply:
It seems rare that you would have a newer Yosys than Nexpnr. The other way around would just have slightly worse results.

Reply:
Sure, there is always workaround. Anyway agree that current code is more user prone to errors.

Reply:
So what do you think we should do?
* merge as is
* remove those lines
* remove entire `-json` option (it's kinda useless without these lines)
* rename `-json` to... `-apicula` or something to clarify it does something special

Reply:
Think most reasonable option for now is to merge code as is. If someone have older yosys and newer nextpnr it will not be a problems but will just get not that good result.